### ISSUE TITLE
Invalidations can cause corrupted reads

### ISSUE THREAD

**Issue by stagraqubole** [qubole/rubix#372]
In most cases invalidations are safe given file does not really get deleted as long as there is fd open on it but in following case we can have corrupted read and cause query failures:
```
If reads R happens on a file while a eviction is in progress on the same file then this can arise:

1. R.getCacheStatus comes before the deletions happen, reads stale mdfile state
2. Deletions happen, both mdfile and data file deleted
3. Another read comes in and creates new mdfile and new data file
4. Now CachedRRC for ‘R’ opens fileChannel and reads based on the stale mdfile response
5. fileChannel will allow reading from new file beyond the areas that have been written to hence will read holes 
```

This is going to be solved by using generation-number for cached files, details [here](https://docs.google.com/document/d/1VlK162PGDl8DbevRCgWNItU1ZJulqATHcGHe-ybBGo4/edit)

**Comment by sopel39** [comment]
@stagraqubole are you actively working on this?

**Comment by raunaqmorarka** [comment]
@sopel39 There is an open PR to address this https://github.com/qubole/rubix/pull/381

**Comment by stagraqubole** [comment]
Fixed by #381
